msprime/stats.py

```diff
@@ -711,3 +720,155 @@ def wfn(x):
+        assert mutations.site[0] == 0
+        # index of *left-hand* end of the current window
+        window_num = 0
+        for interval, records_out, records_in in self.tree_sequence.edge_diffs():
+            if site_num == nsites:
+                break
+            for sign, records in ((-1, records_out), (+1, records_in)):
+                for edge in records:
+                    # print("Record (",sign,"):",node,children,time)
+                    # print("\t",X, "-->", L)
+                    dx = [0 for k in range(num_sample_sets)]
+                    if sign == +1:
+                        pi[edge.child] = edge.parent
+                    for k in range(num_sample_sets):
+                        dx[k] += sign * X[edge.child][k]
+                    if sign == -1:
+                        pi[edge.child] = -1
+                    for k in range(num_sample_sets):
+                        X[edge.parent][k] += dx[k]
+                    # propagate change up the tree
+                    u = pi[edge.parent]
+                    if u != -1:
+                        next_u = pi[u]
+                        while u != -1:
+                            for k in range(num_sample_sets):
+                                X[u][k] += dx[k]
+                            u = next_u
+                            next_u = pi[next_u]
+                    # print("\t",X, "-->", L)
+            # print("next tree:",L,length)
+            site_pos = sites.position[site_num]
+            start_mut = next_mut
+            assert mutations.site[start_mut] == site_num
+            while site_pos < interval[1]:
+                if site_pos > windows[window_num + 1]:
+                    window_length = windows[window_num + 1] - windows[window_num]
+                    for j in range(n_out):
+                        S[window_num][j] /= window_length
+                    window_num += 1
+                mut_nodes = [-1] # -1 will stand in for the root
+                U = [n]  # number of nonmutated samples below node
+                alleles = [sites.ancestral_state[site_num]]
+                parent_muts = [-1] # index in mut_nodes of the parent mut
+                while ((next_mut < nmuts) and
+                       (mutations.site[next_mut] == site_num)):
+                    m = mutations.node[next_mut]
+                    mut_nodes.append(m)
+                    U.append(X[m])
+                    alleles.append(mutations.derived_state[next_mut])
+                    p = mutations.parent[next_mut]
+                    if p != -1:
+                        parent_muts.append(p - start_mut + 1)
+                    else:
+                        parent_muts.append(0)
+                    next_mut += 1
+                nm = next_mut - start_mut
+                for j in range(nm):
+                    if parent_muts[j] != -1:
+                        for k in range(num_sample_sets):
+                            U[parent_muts[j]][k] -= X[mut_nodes[j]][k]
```

Comment:
Here's the place where we use the "mutation parent" information. @jeromekelleher